File: algorithm/25.cpp
//每隔K个数反转一次链表，leetcode原题：lc25

#include <iostream>
using namespace std;

struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int v=0):val(v), next(nullptr) {}
};

class Solution
{
public:
    ListNode* reverseKGroup(ListNode* head, int k);
};

ListNode* Solution::reverseKGroup(ListNode* head, int k)
{
    if(head==nullptr || head->next==nullptr) return head;
    
    ListNode* dummyHead = new ListNode(-1);
    dummyHead->next = head;
    
    ListNode* prev = nullptr;
    ListNode* cur = head;
    ListNode* next = nullptr;
    ListNode* prevLast = dummyHead;
    ListNode* curFirst = head;
    int count = 0;
    
    while(cur!=nullptr)
    {
        if(count==0)
        {
            curFirst = cur;
        }
        count++;
        next = cur->next;
        cur->next = prev;
        prev = cur;
        cur = next;
        if(count==k)
        {
            prevLast->next = prev;
            prevLast = curFirst;
            prev = nullptr;
            count = 0;
        }
    }

    if(count<k) prevLast->next = reverseKGroup(prev, count);
    
    head = dummyHead->next;
    delete dummyHead;
    return head;
}

int main() {
    ListNode* n1 = new ListNode(1);
    ListNode* n2 = new ListNode(2);
    ListNode* n3 = new ListNode(3);
    ListNode* n4 = new ListNode(4);
    ListNode* n5 = new ListNode(5);
    n1->next = n2;
    n2->next = n3;
    n3->next = n4;
    n4->next = n5;

    int k = 2;

    Solution obj;
    ListNode* result = obj.reverseKGroup(n1, k);

    return 0;
}

File: algorithm/go/firstround/25.go
package firstround

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil || k < 2 {
		return head
	}

	dummy := &ListNode{Next: head}
	prev := dummy
	cur := head
	for cur != nil {
		node := cur
		// 判断剩余节点数量是否大于等于k，小于k则无需反转
		for i := 0; i < k; i++ {
			if cur == nil {
				return dummy.Next
			}
			cur = cur.Next
		}

		// 反转从node开始的k个节点
		h := node
		var p *ListNode
		for i := 0; i < k; i++ {
			next := node.Next
			node.Next = p
			p = node
			node = next
		}
		prev.Next = p
		h.Next = cur

		// 反转完成之后更新prev
		prev = h
	}

	return dummy.Next
}
